File: src/agents/alerter/src/farmos/commit-db.js
'use strict';

// Phase 40 D-02 / D-02a / D-07: farmOS-write commit lifecycle persistence.
//
// Extends Phase 38/39's signal_draft table with the columns the commit
// pipeline needs and exposes the SQL primitives the commit-watchdog
// (Plan 05) and integration tests (Plan 07) consume.
//
// Allowed signal_draft.status values after this migration (validated in JS,
// NOT via pg CHECK constraint -- mirrors Phase 38/39 precedent):
//   pending | awaiting_farmer | confirmed | discarded | expired |
//   needs_review | committing | committed | commit_failed
//
// Atomic transitions all use a conditional UPDATE WHERE status='<prev>' so
// duplicate calls + watchdog/restart races collapse to rowCount=0 no-ops.
// All write helpers never-throw (try/catch returning {ok:false, reason}).
// No em-dashes anywhere in this file (memory: feedback_no_em_dashes_in_artifacts).

async function initDb(pool) {
  // Phase 62 D-01: the origin guard column must exist before findConfirmedCandidates
  // runs its `AND origin != 'python'` clause, otherwise the SELECT throws and the
  // never-throws wrapper returns [] (prod commits silently freeze). Adding it here
  // makes the patched alerter self-sufficient on a clean DB; the Python migration
  // adds the identical column independently (idempotent ADD COLUMN IF NOT EXISTS).
  await pool.query(
    `ALTER TABLE signal_draft ADD COLUMN IF NOT EXISTS origin text NOT NULL DEFAULT 'node'`
  );
  await pool.query(
    `ALTER TABLE signal_draft ADD COLUMN IF NOT EXISTS farmos_response jsonb`
  );
  await pool.query(
    `ALTER TABLE signal_draft ADD COLUMN IF NOT EXISTS committed_at timestamptz`
  );
  await pool.query(
    `ALTER TABLE signal_draft ADD COLUMN IF NOT EXISTS commit_failed_reason text`
  );
  await pool.query(
    `ALTER TABLE signal_draft ADD COLUMN IF NOT EXISTS commit_attempt_count int NOT NULL DEFAULT 0`
  );
  await pool.query(
    `ALTER TABLE signal_draft ADD COLUMN IF NOT EXISTS committed_at_attempt timestamptz`
  );
  // Phase 45 D-01 (ACK-04): mark-then-send idempotency claim for commit
  // outcome acks. The CAS UPDATE in tryMarkOutcomeAckSent is the only writer.
  // No index: column is never scanned, only single-row CAS by id.
  await pool.query(
    `ALTER TABLE signal_draft ADD COLUMN IF NOT EXISTS outcome_ack_sent_at timestamptz`
  );
  await pool.query(
    `CREATE INDEX IF NOT EXISTS idx_signal_draft_status_confirmed
       ON signal_draft (status, confirmed_at) WHERE status IN ('confirmed','committing')`
  );
}

async function findConfirmedCandidates(pool, batchCap) {
  try {
    const r = await pool.query(
      `SELECT * FROM signal_draft
        WHERE status='confirmed' AND origin != 'python'
        ORDER BY confirmed_at ASC NULLS LAST
        LIMIT $1`,
      [batchCap]
    );
    return r.rows || [];
  } catch (e) {
    return [];
  }
}

async function acquireCommitLock(pool, draftId) {
  try {
    const r = await pool.query(
      `UPDATE signal_draft
          SET status='committing',
              committed_at_attempt = now(),
              commit_attempt_count = commit_attempt_count + 1
        WHERE id=$1 AND status='confirmed'
        RETURNING *`,
      [draftId]
    );
    const row = r.rows && r.rows.length > 0 ? r.rows[0] : null;
    return { ok: true, rowCount: r.rowCount, row };
  } catch (e) {
    return { ok: false, reason: e.message };
  }
}

async function markCommitted(pool, draftId, farmosResponse) {
  try {
    const r = await pool.query(
      `UPDATE signal_draft
          SET status='committed',
              farmos_response = $2::jsonb,
              committed_at = now()
        WHERE id=$1 AND status='committing'`,
      [draftId, farmosResponse == null ? null : JSON.stringify(farmosResponse)]
    );
    return { ok: true, rowCount: r.rowCount };
  } catch (e) {
    return { ok: false, reason: e.message };
  }
}

async function markFailed(pool, draftId, reason) {
  try {
    const r = await pool.query(
      `UPDATE signal_draft
          SET status='commit_failed',
              commit_failed_reason = $2,
              committed_at = now()
        WHERE id=$1 AND status='committing'`,
      [draftId, reason == null ? null : String(reason)]
    );
    return { ok: true, rowCount: r.rowCount };
  } catch (e) {
    return { ok: false, reason: e.message };
  }
}

async function requeueForRetry(pool, draftId) {
  // NOTE: committed_at_attempt is PRESERVED across requeue so the watchdog's
  // pre-lock backoff gate (Plan 05 task 2) can compare clock.now() - prev to
  // the configured backoff. releaseStaleLocks (the crash-recovery path) is
  // what NULLs committed_at_attempt back to a clean state.
  try {
    const r = await pool.query(
      `UPDATE signal_draft
          SET status='confirmed'
        WHERE id=$1 AND status='committing'`,
      [draftId]
    );
    return { ok: true, rowCount: r.rowCount };
  } catch (e) {
    return { ok: false, reason: e.message };
  }
}

async function releaseStaleLocks(pool, staleMin) {
  try {
    const r = await pool.query(
      `UPDATE signal_draft
          SET status='confirmed',
              committed_at_attempt = NULL
        WHERE status='committing'
          AND committed_at_attempt < now() - ($1 || ' minutes')::interval
        RETURNING id`,
      [String(staleMin)]
    );
    const released_ids = (r.rows || []).map((row) => row.id);
    return { ok: true, rowCount: r.rowCount, released_ids };
  } catch (e) {
    return { ok: false, reason: e.message, released_ids: [] };
  }
}

async function getCachedResponse(pool, draftId) {
  try {
    const r = await pool.query(
      `SELECT status, farmos_response, commit_failed_reason
         FROM signal_draft WHERE id=$1`,
      [draftId]
    );
    if (!r.rows || r.rows.length === 0) {
      return { ok: false, reason: 'not_found' };
    }
    const row = r.rows[0];
    return {
      ok: true,
      status: row.status,
      farmos_response: row.farmos_response,
      commit_failed_reason: row.commit_failed_reason,
    };
  } catch (e) {
    return { ok: false, reason: e.message };
  }
}

async function getAttemptCount(pool, draftId) {
  try {
    const r = await pool.query(
      `SELECT commit_attempt_count FROM signal_draft WHERE id=$1`,
      [draftId]
    );
    if (!r.rows || r.rows.length === 0) return { ok: false, reason: 'not_found' };
    return { ok: true, commit_attempt_count: r.rows[0].commit_attempt_count };
  } catch (e) {
    return { ok: false, reason: e.message };
  }
}

// Phase 45 D-01 (ACK-04): mark-then-send claim primitive.
//
// Single-statement CAS that converts a NULL outcome_ack_sent_at into now() and
// returns the row exactly once. Second + Nth callers see rowCount=0 and get
// { ok:false, reason:'already_claimed' }. Unknown id => { ok:false,
// reason:'not_found' }. SQL errors throw (consistent with the read helpers'
// shape, distinct from the never-throw write helpers above; this is the
// dispatch-side primitive and callers must see hard failures).
async function tryMarkOutcomeAckSent(pool, draftId) {
  const exists = await pool.query(
    `SELECT 1 FROM signal_draft WHERE id=$1`,
    [draftId]
  );
  if (!exists.rows || exists.rows.length === 0) {
    return { ok: false, reason: 'not_found' };
  }
  const r = await pool.query(
    `UPDATE signal_draft
        SET outcome_ack_sent_at = now()
      WHERE id=$1 AND outcome_ack_sent_at IS NULL
      RETURNING id, outcome_ack_sent_at`,
    [draftId]
  );
  if (r.rowCount === 0) {
    return { ok: false, reason: 'already_claimed' };
  }
  return { ok: true, id: r.rows[0].id, claimed_at: r.rows[0].outcome_ack_sent_at };
}

module.exports = {
  initDb,
  findConfirmedCandidates,
  acquireCommitLock,
  markCommitted,
  markFailed,
  requeueForRetry,
  releaseStaleLocks,
  getCachedResponse,
  getAttemptCount,
  tryMarkOutcomeAckSent,
};

File: src/farm-agent/farm_agent/farmos/commit_db.py
"""farm_agent/farmos/commit_db.py -- Python commit-lifecycle DAO for signal_draft.

Faithful Python port of src/agents/alerter/src/farmos/commit-db.js
(Phase 40 D-02 / Phase 62-10).

ORIGIN GUARD (Phase 62 D-01 / T-62-27):
  The live Node watchdog SELECTs `WHERE status='confirmed' AND origin != 'python'`
  to avoid draining Python-owned rows. This Python side is the mirror:
    - find_confirmed_candidates: SELECT WHERE status='confirmed' AND origin='python'
    - acquire_commit_lock, mark_committed, requeue_for_retry: SET origin='python'
  Any Python-touched row stays Python-owned and the live Node watchdog ignores it.

Write helpers are never-throws (try/except Exception) returning
  {"ok": True, "rowcount": n} | {"ok": False, "reason": str(e)}
find_* queries return [] on error.

SQL port notes:
  - Node $1/$2 -> Python %s/%s (psycopg3 paramstyle)
  - staleMin String cast -> str(stale_min) with interval-as-string pattern
  - JSON serialized via json.dumps (None -> NULL)
  - No CHECK constraint reliance -- statuses validated in application code

No em-dashes in source artifacts.
"""
from __future__ import annotations

import json
import logging

from psycopg_pool import AsyncConnectionPool

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# SQL constants
# ---------------------------------------------------------------------------

# Phase 62 D-01: origin='python' guard in SELECT.
# The Node watchdog selects WHERE origin != 'python'; Python side selects WHERE origin='python'.
_FIND_CONFIRMED_SQL = """
SELECT * FROM signal_draft
 WHERE status='confirmed'
   AND origin='python'
 ORDER BY confirmed_at ASC NULLS LAST
 LIMIT %s
"""

_ACQUIRE_LOCK_SQL = """
UPDATE signal_draft
   SET status='committing',
       committed_at_attempt=NOW(),
       commit_attempt_count=commit_attempt_count + 1,
       origin='python'
 WHERE id=%s AND status='confirmed'
 RETURNING *
"""

_MARK_COMMITTED_SQL = """
UPDATE signal_draft
   SET status='committed',
       farmos_response=%s::jsonb,
       committed_at=NOW(),
       origin='python'
 WHERE id=%s AND status='committing'
"""

_MARK_FAILED_SQL = """
UPDATE signal_draft
   SET status='commit_failed',
       commit_failed_reason=%s,
       commit_failed_transport=%s,
       committed_at=NOW()
 WHERE id=%s AND status='committing'
"""

# NOTE: committed_at_attempt is PRESERVED across requeue so the watchdog
# backoff gate can compare now() - prev. releaseStaleLocks NULLs it back.
_REQUEUE_SQL = """
UPDATE signal_draft
   SET status='confirmed',
       origin='python'
 WHERE id=%s AND status='committing'
"""

_RELEASE_STALE_SQL = """
UPDATE signal_draft
   SET status='confirmed',
       committed_at_attempt=NULL
 WHERE status='committing'
   AND committed_at_attempt < NOW() - (%s || ' minutes')::interval
 RETURNING id
"""

# Phase 62 D-06: hold draft as fidelity_cross_check_unverified (T-62-30 / FWR-03).
# Transition from 'committing' (lock already acquired) to the hold status.
# origin='python' preserved so the Node watchdog does not pick it up.
_HOLD_FIDELITY_SQL = """
UPDATE signal_draft
   SET status='fidelity_cross_check_unverified',
       origin='python'
 WHERE id=%s AND status='committing'
"""


# ---------------------------------------------------------------------------
# Finder queries (return [] on error -- never raise)
# ---------------------------------------------------------------------------


async def find_confirmed_candidates(
    pool: AsyncConnectionPool, batch_cap: int = 10
) -> list[dict]:
    """SELECT confirmed rows owned by Python (origin='python').

    Returns [] on error (never raises).
    psycopg3: conn.execute() returns a cursor; cursor.description gives column names.
    """
    try:
        async with pool.connection() as conn:
            cur = await conn.execute(_FIND_CONFIRMED_SQL, (batch_cap,))
            rows = await cur.fetchall()
        if not rows:
            return []
        col_names = [d.name for d in cur.description]
        return [dict(zip(col_names, row, strict=False)) for row in rows]
    except Exception as e:  # noqa: BLE001
        logger.warning("[commit_db] find_confirmed_candidates failed: %s", e)
        return []


# ---------------------------------------------------------------------------
# Write helpers (never raise; return {ok, rowcount} or {ok, reason})
# ---------------------------------------------------------------------------


async def acquire_commit_lock(pool: AsyncConnectionPool, draft_id: str) -> dict:
    """Atomic CAS: status='confirmed' -> 'committing', SET origin='python'.

    rowcount=1: lock acquired (caller should commit).
    rowcount=0: race lost or row not in 'confirmed' (no-op, safe to ignore).
    """
    try:
        async with pool.connection() as conn:
            result = await conn.execute(_ACQUIRE_LOCK_SQL, (draft_id,))
        return {"ok": True, "rowcount": result.rowcount}
    except Exception as e:  # noqa: BLE001
        logger.warning("[commit_db] acquire_commit_lock failed draft_id=%s: %s", draft_id, e)
        return {"ok": False, "reason": str(e)}


async def mark_committed(
    pool: AsyncConnectionPool, draft_id: str, farmos_response: dict | None
) -> dict:
    """CAS: status='committing' -> 'committed', store farmos_response, SET origin='python'.

    rowcount=0 if not currently 'committing' (idempotent no-op).
    """
    try:
        response_json = json.dumps(farmos_response) if farmos_response is not None else None
        async with pool.connection() as conn:
            result = await conn.execute(_MARK_COMMITTED_SQL, (response_json, draft_id))
        return {"ok": True, "rowcount": result.rowcount}
    except Exception as e:  # noqa: BLE001
        logger.warning("[commit_db] mark_committed failed draft_id=%s: %s", draft_id, e)
        return {"ok": False, "reason": str(e)}


async def mark_failed(
    pool: AsyncConnectionPool, draft_id: str, reason: str | None, transport: bool = False
) -> dict:
    """CAS: status='committing' -> 'commit_failed', store commit_failed_reason.

    rowcount=0 if not currently 'committing' (idempotent no-op).

    MUSHY-75: `transport` records whether the failure was the server being
    unreachable rather than the entry being wrong. It cannot be recovered from
    the reason string later, and the recovery pass needs it.
    """
    try:
        reason_str = str(reason) if reason is not None else None
        async with pool.connection() as conn:
            result = await conn.execute(_MARK_FAILED_SQL, (reason_str, bool(transport), draft_id))
        return {"ok": True, "rowcount": result.rowcount}
    except Exception as e:  # noqa: BLE001
        logger.warning("[commit_db] mark_failed failed draft_id=%s: %s", draft_id, e)
        return {"ok": False, "reason": str(e)}


async def requeue_for_retry(pool: AsyncConnectionPool, draft_id: str) -> dict:
    """CAS: status='committing' -> 'confirmed', SET origin='python'.

    committed_at_attempt is PRESERVED (watchdog backoff gate uses it).
    rowcount=0 if not currently 'committing' (no-op).
    """
    try:
        async with pool.connection() as conn:
            result = await conn.execute(_REQUEUE_SQL, (draft_id,))
        return {"ok": True, "rowcount": result.rowcount}
    except Exception as e:  # noqa: BLE001
        logger.warning("[commit_db] requeue_for_retry failed draft_id=%s: %s", draft_id, e)
        return {"ok": False, "reason": str(e)}


async def hold_for_fidelity(pool: AsyncConnectionPool, draft_id: str) -> dict:
    """CAS: status='committing' -> 'fidelity_cross_check_unverified', SET origin='python'.

    Called by commit_watchdog when the fidelity gate returns a strain_mismatch hold
    (D-06 / T-62-30). The draft is held for human review; the farmOS commit is NOT
    made. origin='python' prevents the Node watchdog from picking up this row.

    rowcount=0 if not currently 'committing' (idempotent no-op).
    """
    try:
        async with pool.connection() as conn:
            result = await conn.execute(_HOLD_FIDELITY_SQL, (draft_id,))
        return {"ok": True, "rowcount": result.rowcount}
    except Exception as e:  # noqa: BLE001
        logger.warning("[commit_db] hold_for_fidelity failed draft_id=%s: %s", draft_id, e)
        return {"ok": False, "reason": str(e)}


async def release_stale_locks(
    pool: AsyncConnectionPool, stale_min: int = 15
) -> dict:
    """Reclaim locks stuck in 'committing' older than stale_min minutes.

    Uses interval-as-string pattern per confirm_repo.py:
      (%s || ' minutes')::interval with str(stale_min)

    Returns {"ok": True, "rowcount": n, "released_ids": [id, ...]}
    on success; {"ok": False, "reason": str, "released_ids": []} on error.
    """
    try:
        async with pool.connection() as conn:
            result = await conn.execute(_RELEASE_STALE_SQL, (str(stale_min),))
            rows = await result.fetchall()
        released_ids = [row[0] for row in (rows or [])]
        return {"ok": True, "rowcount": result.rowcount, "released_ids": released_ids}
    except Exception as e:  # noqa: BLE001
        logger.warning("[commit_db] release_stale_locks failed stale_min=%s: %s", stale_min, e)
        return {"ok": False, "reason": str(e), "released_ids": []}


# ---------------------------------------------------------------------------
# MUSHY-75 -- transport-parked recovery
# ---------------------------------------------------------------------------

_FIND_TRANSPORT_PARKED_SQL = """
SELECT id, log_type, draft_json, sender_e164, commit_attempt_count
  FROM signal_draft
 WHERE status='commit_failed'
   AND commit_failed_transport IS TRUE
   AND origin='python'
 ORDER BY confirmed_at
 LIMIT 20
"""

_REQUEUE_PARKED_SQL = """
UPDATE signal_draft
   SET status='confirmed',
       commit_attempt_count=0,
       commit_failed_reason=NULL,
       commit_failed_transport=NULL
 WHERE id=%s AND status='commit_failed' AND commit_failed_transport IS TRUE
"""


async def find_transport_parked(pool: AsyncConnectionPool) -> list[dict]:
    """Drafts stuck at the attempt cap because farmOS was unreachable.

    Bounded at 20: a recovery pass runs inside the watchdog tick and a farmOS
    outage could park many drafts at once. The rest come back next tick.
    """
    try:
        async with pool.connection() as conn:
            cur = await conn.execute(_FIND_TRANSPORT_PARKED_SQL)
            cols = [d[0] for d in cur.description]
            return [dict(zip(cols, r)) for r in await cur.fetchall()]
    except Exception as e:  # noqa: BLE001
        logger.warning("[commit_db] find_transport_parked failed: %s", e)
        return []


async def requeue_parked(pool: AsyncConnectionPool, draft_id: str) -> dict:
    """CAS: 'commit_failed' (transport) -> 'confirmed', attempts reset to 0.

    The status+transport guard is what stops this resurrecting a validation
    failure, which really does need the farmer to fix it.
    """
    try:
        async with pool.connection() as conn:
            result = await conn.execute(_REQUEUE_PARKED_SQL, (draft_id,))
        return {"ok": True, "rowcount": result.rowcount}
    except Exception as e:  # noqa: BLE001
        logger.warning("[commit_db] requeue_parked failed draft_id=%s: %s", draft_id, e)
        return {"ok": False, "reason": str(e)}
